crates/swc_ecma_utils/src/function/function_wrapper.rs

```diff
@@ -0,0 +1,394 @@
+
+        let assign_stmt = AssignExpr {
+            span: DUMMY_SP,
+            op: AssignOp::Assign,
+            left: PatOrExpr::Expr(Box::new(Expr::Ident(ref_ident.clone()))),
+            right: Box::new(self.function.take()),
+        }
+        .into_stmt();
+
+        // clone `return REF.apply(this, arguments);`
+        let return_ref_apply_stmt = function.body.as_ref().unwrap().stmts[0].clone();
+
+        let ref_fn_block_stmt = BlockStmt {
+            span: DUMMY_SP,
+            stmts: vec![assign_stmt, return_ref_apply_stmt],
+        };
+
+        let ref_decl = FnDecl {
+            ident: ref_ident,
+            declare: false,
+            function: Function {
+                span: DUMMY_SP,
+                is_async: false,
+                is_generator: false,
+                params: self.params.take(),
+                body: Some(ref_fn_block_stmt),
+                decorators: Default::default(),
+                type_params: Default::default(),
+                return_type: Default::default(),
+            },
+        };
+
+        let name_decl = FnDecl {
+            ident: name_ident,
+            function,
+            declare: false,
+        };
+
+        (name_decl, ref_decl)
+    }
+
+    ///
+    /// ```javascript
+    /// function NAME(PARAMS) {
+    ///     return REF.apply(this, arguments);
+    /// }
+    /// ```
+    fn build_function_forward(&mut self, ref_ident: Ident, name_ident: Option<Ident>) -> FnExpr {
+        let apply = Stmt::Return(ReturnStmt {
+            span: DUMMY_SP,
+            arg: Some(Box::new(ref_ident.apply(
+                DUMMY_SP,
+                Box::new(Expr::This(ThisExpr { span: DUMMY_SP })),
+                vec![quote_ident!("arguments").as_arg()],
+            ))),
+        });
+
+        FnExpr {
+            ident: name_ident,
+            function: Function {
+                span: DUMMY_SP,
+                is_async: false,
+                is_generator: false,
+                params: self.params.take(),
+                body: Some(BlockStmt {
+                    span: DUMMY_SP,
+                    stmts: vec![apply],
+                }),
+                decorators: Default::default(),
+                type_params: Default::default(),
+                return_type: Default::default(),
+            },
+        }
+    }
+}
+
+impl Into<Expr> for FunctionWrapper {
+    fn into(mut self) -> Expr {
+        // A function may be called internally recursively if it has a name
+        if let Some(name_ident) = self.function_ident.as_ref().cloned() {
+            self.build_named_expression_wrapper(name_ident)
+        }
+        // An internal recursive call has to refer to the external binding_name,
+        // It is safe to using expression wrapper.
+        else if self.params.len() > 0 || self.binding_ident.is_some() {
+            self.build_anonymous_expression_wrapper()
+        } else {
+            // Optimization: no name, no params
```

Comment:
@kdy1  We can only optimize it when there is no function name and no parameters.

Reply:
maybe functions like
```js
function (a = 10, ...rest) {}
```
could also be optimized?

Reply:
```javascript
(function (a = 10, ...rest) {}).length === 0
```

I will create a test to check it.

Reply:
Optional parameters are not counted in the function parameter length.
So when all are optional parameters, or rest parameters, the function parameter length is 0, and we can safely optimize it.

Reply:
FYI, The `self.params` here are cloned and preprocessed by `Self::get_params`.